### PR TITLE
Touch Fuctions + Tests

### PR DESCRIPTION
**Pull request by rkabrick**

### CODE REVIEW

File: tests/test_global.sh [hunk 1]
```diff
@@ -2,11 +2,14 @@
 
 # Output checking
 ./irpc.py ./examples/simple.irp.c > tests/simple.c
-diff ./tests/gold.simple.c tests/simple.c
+diff -w -b -B ./tests/gold.simple.c tests/simple.c
+
+./irpc.py ./examples/touch.irp.c > tests/touch.c
+diff -w -b -B ./tests/gold.touch.c tests/touch.c
```
Comment (TApplencourt) [inline tests/test_global.sh]: Nice change! Thanks!

**Comment by TApplencourt** [comment]
Ok, I don't understand why ba4d28f is failing...

File: tests/gold.touch.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  if (!a_provided)
+  {
+    provide_a();
+    a_provided = true;
+  }
+
+  a = 20;
+  touch_a();
+  b = a;
+}
```
Comment (TApplencourt) [inline tests/gold.touch.c]: It should have a call to `provide_a()`

File: tests/gold.touch.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+{
+  if (!a_provided)
+  {
+    provide_a();
+    a_provided = true;
+  }
+
+  a = 20;
+  touch_a();
+  b = a;
```
Comment (TApplencourt) [inline tests/gold.touch.c]: It should have a call to `provide_a()`

File: irpc.py [hunk 4]
```diff
@@ -120,18 +178,60 @@ def add_provider_call(funcdef: FuncDef,
...
+
+        for parent_entity, _ in entity2Compound(provdef.body, entnames).items():
+            child_adjacency_graph[parent_entity].add(entity)
 
-    return adjacency_graph
+    return child_adjacency_graph
+
+
+def find_touches(filename):
+    l_touch = set()
```
Comment (TApplencourt) [inline irpc.py:208]: You should recurse the AST and find is we have a funccall `touch_{provider}`

File: irpc.py [hunk 5]
```diff
@@ -120,18 +178,60 @@ def add_provider_call(funcdef: FuncDef,
...
+    with open(filename, 'r') as input:
+        for line in input:
+            if "touch_" in line:
```
Comment (TApplencourt) [inline irpc.py:211]: Indeed, You can have false positive like
```
int my_variable_touch_a = 10
```
This is not a fun call, but will pass your test

File: irpc.py [hunk 6]
```diff
@@ -109,6 +103,70 @@ def hoist_declaration(main: FuncDef,
...
+        # If anything but instance of ID -> recurse 
+        else:
+            # If Compound: New compound = node & j (index of inner compound) is initialized to 0
+            # Else: Recurse over same compound as before with index value used for old compound
+            c, j  = (node, 0) if isinstance(node, Compound) else (old_compound, i)
+
+            # Recursive call followed by updating the dictionary
+            for k,v in find_instances(node, l_ent, c, j).items():
+                d[k] += v
+
```
Comment (TApplencourt) [inline irpc.py]: Nice comments! Thanks!

File: examples/touch.irp.c [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+void provide_b(){
+  a = 20;
+  touch_a();
+  int b;
+  b = a;
+}
+
+int main(){
+  printf("b = %i", b);
+}
```
Comment (TApplencourt) [inline examples/touch.irp.c:15]: May we should do A -> B -> C where C touch A and use b?
So we will need to call `provide_b` after the `touch_a()`, and our test will be more robust (testing more things).
Or we can have two tests. More test is better :)

File: tests/gold.touch.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  if (!b_provided)
+  {
+    provide_b();
+    b_provided = true;
+  }
+
+  printf("b = %i", b);
+}
+
+
```
Comment (TApplencourt) [inline tests/gold.touch.c:41]: Does this thing compile? We use `touch_a()` before declaring it.

**Comment by TApplencourt** [comment]
> I don't understand why this is failing either... The error message in the log makes no sense especially with the included flags. Also, the test passes every time I run it locally

I guess it's something to do with the randomness of `set`. But yes, it strange. I will investigate more...

File: tests/test_global.sh [hunk 9]
```diff
@@ -2,11 +2,14 @@
...
```
Comment (TApplencourt) [inline tests/test_global.sh]: Please add a test for the compilation of the `touch.c`

Also, You can split test_global in `test_simple` and `test_touch`. Maybe more readable

File: tests/test_global.sh [hunk 10]
```diff
@@ -2,11 +2,14 @@
...
 
 # Compiling to be sure
 gcc tests/simple.c -o tests/simple.exe
 ./tests/simple.exe > tests/simple.out
 
 # Output checking
-diff ./tests/gold.simple.out tests/simple.out
+diff -w -b -B ./tests/gold.simple.out tests/simple.out
```
Comment (TApplencourt) [inline tests/test_global.sh]: Nice change! Thanks!

File: tests/gold.touch.c (hunk 8 above)
Comment (rkabrick) [reply tests/gold.touch.c:41]: Yes... you had mentioned you didn't want the user to have to declare the touch functions so as of now, it works like that.

**Pull request closed without merging**